content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+Now we have a (pretty big) JSON file with lots of information in it! To inspect any JSON data, I like to use [jq][]. To have an idea of what the data looks like, get the first five top objects from the JSON's `result` field:
+
+```
+= jq '.result[0:5]' server-test-1_dlv-rpc-list_goroutines.json
+[
+  {
+    "id": 1,
+    "currentLoc": {
+      "pc": 4387627,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 303,
+      "function": {
+        "name": "runtime.gopark",
+        "value": 4387392,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "userCurrentLoc": {
+      "pc": 7351940,
+      "file": "/vagrant/example/server/main.go",
+      "line": 56,
+      "function": {
+        "name": "main.run",
+        "value": 7351152,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "goStatementLoc": {
+      "pc": 4566329,
+      "file": "/usr/local/go/src/runtime/asm_amd64.s",
+      "line": 201,
+      "function": {
+        "name": "runtime.rt0_go",
+        "value": 4566032,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "startLoc": {
+      "pc": 4386096,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 110,
+      "function": {
+        "name": "runtime.main",
+        "value": 4386096,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "threadID": 0
+  },
+  {
+    "id": 2,
+    "currentLoc": {
+      "pc": 4387627,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 303,
+      "function": {
+        "name": "runtime.gopark",
+        "value": 4387392,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "userCurrentLoc": {
+      "pc": 4387627,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 303,
+      "function": {
+        "name": "runtime.gopark",
+        "value": 4387392,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "goStatementLoc": {
+      "pc": 4387029,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 240,
+      "function": {
+        "name": "runtime.init.4",
+        "value": 4386976,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "startLoc": {
+      "pc": 4387056,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 243,
+      "function": {
+        "name": "runtime.forcegchelper",
+        "value": 4387056,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "threadID": 0
+  },
+  ...
+```
+
+Every object in the JSON represents a single goroutine. By checking [“goroutines” command help](https://github.com/derekparker/delve/blob/master/Documentation/cli/README.md#goroutines) we can figure out what data Delve knows. We're interested in `userCurrentLoc` field, which is "topmost stackframe in user code".
```

Comment:
The second sentence is a bit odd, and leaves the reader wondering what the "topmost stackframe in user code" is. My suggestion:
```suggestion
Every object in the JSON represents a single goroutine. The manual for Delev's [“goroutines” command help](https://github.com/derekparker/delve/blob/master/Documentation/cli/README.md#goroutines) tells us what Delve knows about each goroutine. We're interested in the field `userCurrentLoc`. Its description says it is the "topmost stackframe in user code", which means that this is usually the last location in our code the goroutine came across.
```